### ISSUE TITLE
find_homography_dlt - possible improvement and usage

### ISSUE THREAD

**Issue by dkoguciuk** [kornia/kornia#690]
## ❓ Usage

Hi, Kornia community. Did anyone use `find_homography_dlt` as a part of an end-to-end network? I am working on end-to-end image registration project and tried to incorporate it into the network architecture, without any success so far. I've set a small learning rate (i.e. 1e-6/8/10/12) and after a couple of iterations weights of my network explode - what's interesting, gradients do not explode... I am having a hard time understanding what's wrong here - maybe the whole idea of inserting `find_homography_dlt` in end-to-end fashion is wrong?

Any tips? 

## ❓ Possible improvement  

I could have a couple of possible improvements over the current version, let me start with this one: https://github.com/kornia/kornia/blob/master/kornia/geometry/homography.py#L39 , where we have a manual iteration over all points.

```
    w_list = []
    axy_list = []
    for i in range(points1.shape[1]):
        axy_list.append(ax[:, i])
        axy_list.append(ay[:, i])
        w_list.append(weights[:, i])
        w_list.append(weights[:, i])
    A = torch.stack(axy_list, dim=1)
    w = torch.stack(w_list, dim=1)
```

Is there any reason why we do this manually? Why don't we use:

```
A = torch.cat((ax, ay), dim=-1).reshape(ax.shape[0], -1, ax.shape[-1])
w = weights.repeat(1, 2)
```

**Comment by ducha-aiki** [comment]
Hi,

Welcome here and thanks for the reporting!

>  I am having a hard time understanding what's wrong here - maybe the whole idea of inserting find_homography_dlt in end-to-end fashion is wrong?

That is interesting problem. I understand that you probably could not share code, but may be you could create some minimal exploding example on tiny dataset - then I can look into and try to understand what's going on.

> Is there any reason why we do this manually? Why don't we use:

No particular reason, feel free to propose PR

**Comment by edgarriba** [comment]
@dkoguciuk does it make any speed performance improvement ? we recently started to ask to provide small colabs to justify potential changes to improve current functionalities. You can check an example here: https://colab.research.google.com/drive/1dVHXicxHVPfUdyuYNdd4jOzx7JKtwqxb?usp=sharing

**Comment by dkoguciuk** [comment]
Hi @ducha-aiki , Hi @edgarriba ,

thanks for the replies. 

## find_homography_dlt usage

I was wrong. It looks like using `find_homography_dlt` always gives 0 gradients. Maybe I've messed something up, or maybe `find_homography_dlt` does not produce valid gradients. Did you check it?

I've prepared a minimal example to reproduce the error I got:  [HERE](https://github.com/dkoguciuk/kornia-homography-backprop/blob/master/kornia_homography_backprop.ipynb). The whole idea is simple, but unfortunately, I had to produce moderately big amounts of code for that. The main idea is to use `find_homography_dlt` on top of the perspective field produced by CNN (perspective field idea is inspired by [Rui Zeng et al.](https://eprints.qut.edu.au/126933/). Please let me know if I should clarify something.

## possible improvements

When using a big number of points it makes a huge difference. I will provide collab comparison and possibly PR for that. One more possible improvement is to return blind homography on the same device as the input data https://github.com/kornia/kornia/blob/master/kornia/geometry/homography.py#L55

```return torch.empty(points1_norm.size(0), 3, 3)```

Could be fixed with:

```return torch.empty((points1_norm.size(0), 3, 3), dtype=torch.float32, device=points1.device)```

**Comment by ducha-aiki** [comment]
Hi,

Thanks! First, I have modified a bit your example here:

https://github.com/ducha-aiki/kornia-homography-backprop/blob/master/kornia_homography_backprop_single_example.ipynb

Specifically, overfit the single example 

`for step in range(50):
    
    # Prepare ground truth
    #image_warped_np, corners_np, delta_gt_np, hom_gt_np = prepare_train_sample()
    image_warped_np, corners_np, delta_gt_np, hom_gt_np = prepare_train_sample(delta_np = np.array([[8, 8], [-8, 8], [-8, -8], [8, -8]]),
                                                       visualize=False)
`

and changed your network, as BN+ReLU for the output layer cannot in principle produce correct optical flow. And replace BN with GN, as BN for batch_size = 1 is very unstable.

        self.layer5 = nn.Sequential(nn.Conv2d(in_channels=128, out_channels=128, kernel_size=3, padding=1, stride=1, bias=False),
                                    #nn.BatchNorm2d(128),
                                    # BN does not work well for BS =1. Let's replace with group norm
                                    nn.GroupNorm(4, 128),
                                    nn.ReLU())
        self.layer6 = nn.Sequential(nn.Conv2d(in_channels=128, out_channels=2, kernel_size=3, padding=1, stride=1, bias=True),
                                   # nn.BatchNorm2d(2),
                                   # nn.ReLU()
                                    #ReLU for the output does not make sense, because field can be negative
                                    # And the same for the batchnorm - output can be huge and BN does not allow that
                                   )

It successfully overfits the single example.

But the random (as in your example) still fails. 
https://github.com/ducha-aiki/kornia-homography-backprop/blob/master/kornia_homography_backprop-fix.ipynb

The reason is that SVD fails 

        # apply weights
    w_diag = torch.diag_embed(w)
    A = A.transpose(-2, -1) @ w_diag @ A
    try:
        U, S, V = torch.svd(A)
    except:
        print ('SVD Fails!')
        return torch.empty(points1_norm.size(0), 3, 3)

![image](https://user-images.githubusercontent.com/4803565/94266081-e47f7b00-ff39-11ea-846d-5edb5002256c.png)


I don't have solution yet, but at least we should throw a warning for user.
cc: @edgarriba

**Comment by edgarriba** [comment]
might be the case that the input matrix in svd is ill conditioned similar to what we found [here](https://github.com/kornia/kornia/blob/master/kornia/geometry/subpix/spatial_soft_argmax.py#L604) ?

**Comment by dkoguciuk** [comment]
Hi @ducha-aiki ,

nice insights. I wasn't aware of that: `BN+ReLU for the output layer cannot in principle produce correct optical flow`, but if I think about it, you're right, it does make sense. Is it some urban knowledge? This is only toy model I've build to keep it simple, but thanks anyway.

I believe my biggest mistake was putting calculation of norm of gradients before loss.backward(), which gave me always 0 gradients. Clipping gradients indeed helps. Thank you @ducha-aiki!

@edgarriba - it is very likely that the matrix is illposed. Do I understand correctly that one of the possible fixes is to add some epsilon noise to the matrix? 

And lastely, I've never used collab, could you please tell me how to start one to compare possible improvements?

**Comment by edgarriba** [comment]
@dkoguciuk in fact. this should be the proper solution: https://www.reddit.com/r/compsci/comments/3chxty/what_should_i_do_with_an_illconditioned_matrix/

A for colab, we can use this one to find what values makes svd to crash
https://colab.research.google.com/drive/11np62TTbNnvFRJKvJUB5zTTyoaN8FfMu?usp=sharing

I don't know why random noise in that case solves the problem. /cc @ducha-aiki

**Comment by ducha-aiki** [comment]
>Is it some urban knowledge? 

@dkoguciuk no, just from the fact that displacements == optical flow in your example was a)negative -> do not survive ReLU and big, e.g. 8,therefore would be scaled down by batchnorm. I am glad, that I helped :)

>I don't know why random noise in that case solves the problem

@edgarriba 
Because for that particular case, it not that matrix was somehow singular, it was, in fact all zeros or all ones. That is why noise helped. Not sure for the general case...

**Comment by dkoguciuk** [comment]
@ducha-aiki , @edgarriba 

I've edited the [collab](https://colab.research.google.com/drive/11np62TTbNnvFRJKvJUB5zTTyoaN8FfMu) that @edgarriba shared. There are two versions of the DLT function:

1) _vectorized - with the `A` and `w` creation I've mentioned in the first post.
2) _optional - with weights being an optional argument 

Do you find those changes useful?

**Comment by edgarriba** [comment]
@dkoguciuk please go ahead with a PR. As discussed in slack, consider this piece of code:
```python
if weights is None:
      A = A.transpose(-2, -1) @ A
else:
      assert len(weigths.shape) == 1 and weights.shape[0] == A.shape[0], weights.shape
      A = A.transpose(-2, -1) @ A
      # We should use weights
      w_diag = torch.diag_embed(weights.repeat(1, 2))
      A = A.transpose(-2, -1) @ w_diag @ A
```

**Comment by dkoguciuk** [comment]
Merged in #693
